README.md
A Flutter package to get country information in various standards. Data was gathered from www.statoids.com/wab.html

## Features
* [x] Names, ISO 3166-1 alpha-2. alpha-3, numeric, ITU, GEC, IOC, FIFA, DS, WMO, GAUL, MARC and dial country codes
* [x] Country independence status based on the CIA World Factbook
* [x] Adding own countries


## Usage
There are two main ways to use this package.

### 1. Default by using device region
This will allow you to fetch the region of the underlying platform and display the data accordingly.

Assuming an **en-US** region based revice.
```
await CountryCodes.init(); // Optionally, you may provide a `Locale` to get countrie's localizadName

final Locale deviceLocale = CountryCodes.getDeviceLocale();
print(deviceLocale.languageCode); // Displays en
print(deviceLocale.countryCode); // Displays US

final Country details = CountryCodes.detailsForLocale();
print(details.name); // Displays the extended name, for example United States.
print(details.localizedName); // Displays the extended name based on device's language (or other, if provided on init)
print(details.a2); // Displays alpha2Code, for example US.
print(details.a3); // Displays alpha3Code, for example USA.
print(details.dial); // Displays the dial code, for example 1.
print(details.fifa); // Displays FIFA code, for example USA.
print(details.numeric); // Displays numeric code, for example 840.
print(details.gec); // Displays GEC code, for example US.
print(details.itu); // Displays ITU code, for example USA.
print(details.ioc); // Displays IOC code, for example USA.
print(details.ds); // Displays DS code, for example USA.
print(details.wmo); // Displays WMO code, for example US.
print(details.gaul); // Displays GAUL code, for example 259.
print(details.marc); // Displays MARC code, for example xxu.
```
### 2. Use a custom `Locale`
This will use the provided `Locale`, which may not be related to the device's region but instead to the app supported languages.

For example, if your device is on US region but the app only supports PT, you'll get the following:
```
final Country details = CountryCodes.detailsForLocale(Localization.localeOf(context));

print(details.name); // Displays the extended name, for example Portugal.
print(details.a2); // Displays alpha2Code, for example PT.
print(details.a3); // Displays alpha3Code, for example PRT.
print(details.dial); // Displays the dial code, for example 620.
print(details.fifa); // Displays FIFA code, for example POR.
print(details.numeric); // Displays numeric code, for example 840.
print(details.gec); // Displays GEC code, for example PO.
print(details.itu); // Displays ITU code, for example POR.
print(details.ioc); // Displays IOC code, for example POR.
print(details.ds); // Displays DS code, for example P.
print(details.wmo); // Displays WMO code, for example PO.
print(details.gaul); // Displays GAUL code, for example 199.
print(details.marc); // Displays MARC code, for example po.
```

## Credit

A special thank you to miguelruivo.com for his [country_codes](https://pub.dev/packages/country_codes) package. This package have awesome translate solution. And thanks to ersel for JS module [Country Code Info](https://github.com/ersel/country-code-info). This package contains almost every country for this module.

CHANGELOG.md
## [0.0.7] - upgrade gradle versions

## [0.0.6] - quickfix

## [0.0.5] - fix alpha search and change dial return

## [0.0.4] - fix entry file error and fix format

## [0.0.3] - Add native implementation

## [0.0.2] - Update readme and pubspec

## [0.0.1] - Initial release


ios/Classes/SwiftCountryCodesInfoPlugin.swift
import Flutter
import UIKit

public class SwiftCountryCodesInfoPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "country_codes_info", binaryMessenger: registrar.messenger())
    let instance = SwiftCountryCodesInfoPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
        case "getRegion":
            result(Locale.current.regionCode)
            break
        case "getLanguage":
            result(Locale.current.languageCode)
            break
        case "getLocale":
            result([Locale.current.languageCode as Any, Locale.current.regionCode as Any, getLocalizedCountryNames(localeTag: call.arguments as? String)])
            break
        default:
            result(FlutterMethodNotImplemented);
        }
  }

  func getLocalizedCountryNames(localeTag: String?) -> Dictionary<String,String> {
          var localizedCountries:Dictionary<String,String> = [String: String]()

          for countryCode in NSLocale.isoCountryCodes {
              let countryName: String? = NSLocale(localeIdentifier: localeTag ?? Locale.preferredLanguages[0]).displayName(forKey: .countryCode, value: countryCode)
              localizedCountries[countryCode.uppercased()] = countryName ?? "";
          }
          return localizedCountries
      }
}


android/src/main/kotlin/ch/tisagroup/country_codes_info/CountryCodesInfoPlugin.kt
package ch.tisagroup.country_codes_info

import android.R
import java.util.Locale
import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar

/** CountryCodesInfoPlugin */
class CountryCodesInfoPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "country_codes_info")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "getLocale" -> result.success(listOf(Locale.getDefault().language, Locale.getDefault().country, getLocalizedCountryNames(call.arguments as String?)))
      "getRegion" -> result.success(Locale.getDefault().country)
      "getLanguage" -> result.success(Locale.getDefault().language)
      else -> result.notImplemented()
    }
  }

  private fun getLocalizedCountryNames(localeTag: String?) : HashMap<String, String> {
    var localizedCountries: HashMap<String, String> = HashMap()

    val deviceCountry: String = Locale.getDefault().toLanguageTag();

    for (countryCode in Locale.getISOCountries()) {
      val locale = Locale(localeTag ?: deviceCountry,countryCode)
      var countryName: String? = locale.getDisplayCountry(Locale.forLanguageTag(localeTag ?: deviceCountry))
      localizedCountries[countryCode.toUpperCase()] = countryName ?: "";
    }
    return localizedCountries
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}


lib/src/country_codes_info.dart
import 'dart:ui';

import 'package:flutter/services.dart';

import 'countries.dart';

class CountryCodes {
  static const MethodChannel _channel = MethodChannel('country_codes_info');
  static Locale? _deviceLocale;
  static late Map<String, String> _localizedCountryNames;

  static List<Country> get allCountries => Countries.countries;

  static String? _resolveLocale(Locale? locale) {
    locale ??= _deviceLocale;
    assert(locale != null && locale.countryCode != null, '''
         Locale and country code cannot be null. If you are using an iOS simulator, please, make sure you go to region settings and select any country (even if it's already selected) because otherwise your country might be null.
         If you didn't provide one, please make sure you call init before using Country Details
        ''');
    return locale!.countryCode;
  }

  /// Inits the underlying plugin channel and fetch current's device locale to be ready
  /// to use synchronously when required.
  ///
  /// If you never plan to provide a `locale` directly, you must call and await this
  /// by calling `await CountryCodes.init();` before accessing any other method.
  ///
  /// Optionally, you may want to provide your [appLocale] to access localized
  /// country name (eg. if your app is in English, display Italy instead of Italia).
  ///
  /// Example:
  /// ```dart
  /// CountryCodes.init(Localizations.localeOf(context))
  /// ```
  /// This will default to device's language if none is provided.
  static Future<bool> init({Locale? appLocale}) async {
    final List<dynamic>? locale = List<dynamic>.from(
        await (_channel.invokeMethod('getLocale', appLocale?.toLanguageTag())));
    if (locale != null) {
      _deviceLocale = Locale(locale[0], locale[1]);
      _localizedCountryNames = Map.from(locale[2]);
    }
    return _deviceLocale != null;
  }

  /// Returns the current device's `Locale`
  /// Eg. `Locale('en','US')`
  static Locale? getDeviceLocale() {
    assert(_deviceLocale != null,
        'Please, make sure you call await init() before calling getDeviceLocale()');
    return _deviceLocale;
  }

  /// A list of dial codes for every country
  static List<String?> dialNumbers() {
    return allCountries.map((country) => country.dial).toList();
  }

  /// A list of numeric codes for every country
  static List<String?> numbers() {
    return allCountries.map((country) => country.num).toList();
  }

  /// A list of alpha 2 codes for every country
  static List<String?> allAlpha2() {
    return allCountries.map((country) => country.a2).toList();
  }

  /// A list of alpha 3 codes for every country
  static List<String?> allAlpha3() {
    return allCountries.map((country) => country.a3).toList();
  }

  /// A list of ITU codes for every country
  static List<String?> allITU() {
    return allCountries.map((country) => country.itu).toList();
  }

  /// A list of GEC codes for every country
  static List<String?> allGEC() {
    return allCountries.map((country) => country.gec).toList();
  }

  /// A list of IOC codes for every country
  static List<String?> allIOC() {
    return allCountries.map((country) => country.ioc).toList();
  }

  /// A list of FIFA codes for every country
  static List<String?> allFIFA() {
    return allCountries.map((country) => country.fifa).toList();
  }

  /// A list of DS codes for every country
  static List<String?> allDS() {
    return allCountries.map((country) => country.ds).toList();
  }

  /// A list of WMO codes for every country
  static List<String?> allWMO() {
    return allCountries.map((country) => country.wmo).toList();
  }

  /// A list of GAUL codes for every country
  static List<String?> allGAUL() {
    return allCountries.map((country) => country.gaul).toList();
  }

  /// A list of MARC codes for every country
  static List<String?> allMARC() {
    return allCountries.map((country) => country.marc).toList();
  }

  /// A list of country data for every country
  static List<Country?> countries() {
    return allCountries;
  }

  /// Returns the `CountryDetails` for the given [locale]. If not provided,
  /// the device's locale will be used instead.
  /// Have in mind that this is different than specifying `supportedLocale`s
  /// on your app.
  /// Exposed properties are the `name`, 'a2', 'a3', 'num', 'itu', 'gec', 'ioc', 'fifa', 'ds', 'wmo', 'gaul', 'marc', 'dial' and 'independent'
  ///
  /// Example:
  /// ```dart
  /// "name": "United States",
  /// "alpha2Code": "US",
  /// "dial_code": "+1",
  /// ```
  static Country detailsFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return Country.withLocalization(
        country!, _localizedCountryNames[country.a2]);
  }

  static Country? detailsFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return Country.withLocalization(
        country!, _localizedCountryNames[country.a2]);
  }

  /// Returns the ISO 3166-1 `alpha2Code` for the given [locale].
  /// If not provided, device's locale will be used instead.
  /// You can read more about ISO 3166-1 codes [here](https://en.wikipedia.org/wiki/ISO_3166-1)
  ///
  /// Example: (`US`, `PT`, etc.)
  static String? alpha2Code({Locale? locale}) {
    return _resolveLocale(locale);
  }

  /// Returns the ISO 3166-1 `alpha3code` for the given [locale].
  /// If not provided, device's locale will be used instead.
  /// You can read more about ISO 3166-1 codes [here](https://en.wikipedia.org/wiki/ISO_3166-1)
  ///
  /// Example: (`USA`, `PRT`, etc.)
  static String? alpha3Code({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.a3;
  }

  /// Returns the `dialCode` for the given [locale] or device's locale, if not provided.
  ///
  /// Example: (`1`, `351`, etc.)
  static String? dialCodeFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.dial;
  }

  static String? dialCodeFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return country?.dial;
  }

  /// Returns the extended `name` for the given [locale] or if not provided, device's locale or from alpha code.
  ///
  /// Example: (`United States`, `Portugal`, etc.)
  static String? nameFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.name;
  }

  static String? nameFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return country?.name;
  }

  /// Returns the 'fifa' code for the given [locale] or if not provided, device's locale or from alpha code.
  ///
  /// Example: (`USA`, `POR`, etc.)
  static String? fifaFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.fifa;
  }

  static String? fifaFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return country?.fifa;
  }

  /// Returns the numeric code for the given [locale] or if not provided, device's locale or from alpha code.
  ///
  /// Example: (`840`, `620`, etc.)
  static String? numericFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.num;
  }

  static String? numericFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return country?.num;
  }

  /// Returns the GEC code for the given [locale] or if not provided, device's locale or from alpha code.
  ///
  /// Example: (`US`, `PO`, etc.)
  static String? gecFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.gec;
  }

  static String? gecFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return country?.gec;
  }

  /// Returns the ITU code for the given [locale] or if not provided, device's locale or from alpha code.
  ///
  /// Example: (`USA`, `POR`, etc.)
  static String? ituFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.itu;
  }

  static String? ituFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return country?.itu;
  }

  /// Returns the IOC code for the given [locale] or if not provided, device's locale or from alpha code.
  ///
  /// Example: (`USA`, `POR`, etc.)
  static String? iocFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.ioc;
  }

  static String? iocFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return country?.ioc;
  }

  /// Returns the DS code for the given [locale] or if not provided, device's locale or from alpha code.
  ///
  /// Example: (`USA`, `P`, etc.)
  static String? dsFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.ds;
  }

  static String? dsFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return country?.ds;
  }

  /// Returns the WMO code for the given [locale] or if not provided, device's locale or from alpha code.
  ///
  /// Example: (`US`, `PO`, etc.)
  static String? wmoFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.wmo;
  }

  static String? wmoFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return country?.wmo;
  }

  /// Returns the GAUL code for the given [locale] or if not provided, device's locale or from alpha code.
  ///
  /// Example: (`259`, `199`, etc.)
  static String? gaulFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.gaul;
  }

  static String? gaulFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return country?.gaul;
  }

  /// Returns the MARC code for the given [locale] or if not provided, device's locale or from alpha code.
  ///
  /// Example: (`xxu`, `po`, etc.)
  static String? marcFromLocale({Locale? locale}) {
    String? code = _resolveLocale(locale);
    Country? country =
        allCountries.firstWhereSafe((country) => country.a2 == code);
    return country?.marc;
  }

  static String? marcFromAlpha({String? alpha2, String? alpha3}) {
    if (alpha2 == null && alpha3 == null) return null;
    Country? country = allCountries.firstWhereSafe((country) =>
        country.a2 == (alpha2 ?? '') || country.a3 == (alpha3 ?? ''));
    return country?.marc;
  }

  /// Adds given [country] to [Countries.countries] for further usage
  static void addCountries({required List<Country> newCountries}) =>
      allCountries.addAll(newCountries);
}

extension SafeList<Element> on List<Element> {
  Element? firstWhereSafe(bool Function(Element) test) {
    try {
      return firstWhere(test);
    } catch (e) {
      return null;
    }
  }
}
